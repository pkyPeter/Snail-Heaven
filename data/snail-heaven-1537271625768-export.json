{
  "listings" : {
    "item1" : {
      "id" : 12321,
      "listing_url" : "www.yahoo.com"
    },
    "item2" : {
      "id" : 12312312,
      "listing_url" : "www.yahoo.com"
    }
  }
}
