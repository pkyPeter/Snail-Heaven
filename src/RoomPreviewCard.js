import React from "react";
import PropTypes from "prop-types";
import lib from "./lib.js";
import googleMap from "./GoogleMap.js";
import airbnb from "./imgs/airbnb.png";
//FontAwesome主程式
import { library } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { fas } from "@fortawesome/free-solid-svg-icons";
//FontAwesome引用圖片
import { faHeart as faRegularHeart, faSave, faThumbsDown, faEnvelope } from "@fortawesome/free-regular-svg-icons";
import { faHeart as faSolidHeart, faListUl, faThLarge, faSquare, faMapMarkedAlt} from "@fortawesome/free-solid-svg-icons";
library.add(faRegularHeart, faSolidHeart, faSave, faListUl, faThLarge, faSquare, faThumbsDown
  ,faEnvelope, faMapMarkedAlt);

class RoomPreviewCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
    };
    this.stopPropagation = this.stopPropagation.bind(this);
    this.getMarkerBounce = this.getMarkerBounce.bind(this);
    this.stopMarkerBounce = this.stopMarkerBounce.bind(this);
  }
  componentDidMount() {

  }
  componentDidUpdate() {
  
  }
  render() {
    let realEstate = this.props.realEstate;
    let monthly_price = realEstate.monthly_price.toLocaleString("en");
    let loveListStatusIndex = this.props.loveListStatus != null && this.props.getloveListStatusIndex(realEstate.id, this.props.loveListStatus);
    let roomType;
    switch(realEstate.room_type) {
      case "EHA":
      roomType = "整層住家";
      break;
      case "PR":
      roomType = "獨立套房";
      break;
      case "SR":
      roomType = "分租套房";
      break;
    }
    return (
      <div 
      key={this.props.index} 
      className={this.props.resultAreaDisplayType[1]} 
      onClick={(e)=> { this.props.changeSelecteIndex("add",realEstate.index) } } 
      onMouseEnter={(e)=>{ this.getMarkerBounce(e, loveListStatusIndex) }} 
      onMouseLeave={(e)=>{ this.stopMarkerBounce(e, loveListStatusIndex) }}
      >
        <div className="airbnbContainer">
          <img className="airbnb" src={airbnb}></img>
          Airbnb
        </div>
        <div className="img" style={{backgroundImage: `url(${realEstate.picture_url})`}}></div>
        <div className="description">
          <div className="priceGesture absolute">
            <div className="price">{"$" + monthly_price}</div>
            <div className="gesture" onClick={this.stopPropagation}>
              { 
              this.props.loveListStatus[loveListStatusIndex].inList === true 
              ? <FontAwesomeIcon className="icon" icon={["fas","heart"]} style={{ color: "red" }} onClick={(e)=>{ this.props.removeFromLoveList(e, realEstate.id, realEstate); }}/>
              : <FontAwesomeIcon className="icon" icon={["far","heart"]} onClick={(e)=>{ this.props.putIntoLoveList(e, realEstate.id, realEstate); }}/>
              }
              <FontAwesomeIcon className="icon" icon={["far","envelope"]} onClick={()=>{this.props.openEmailForm("",realEstate.id) }}/>
              <FontAwesomeIcon className="icon" icon={["far","thumbs-down"]} onClick={(e)=>{this.props.hideList(e, realEstate.id, realEstate.index);}} />
            </div>
          </div>
          <p>{realEstate.bedrooms} 間房間 · {realEstate.bathrooms} 間廁所 · {roomType}</p>
          <p>{realEstate.district}</p>
          <p className="updateTime">{realEstate.updated}</p>
        </div>
      </div>
    );
  }
  stopPropagation(e) {
    e.stopPropagation();
    e.nativeEvent.stopImmediatePropagation();
  }
  getMarkerBounce(e, index) {
    e.stopPropagation();
    let currentIndex = parseInt(index);
    googleMap.markers[currentIndex].setAnimation(google.maps.Animation.BOUNCE);
  }
  stopMarkerBounce(e, index) {
    let currentIndex = parseInt(index);
    googleMap.markers[currentIndex].setAnimation(null);
  }
}

export default RoomPreviewCard;