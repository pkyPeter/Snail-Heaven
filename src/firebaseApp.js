import firebase from "firebase";
// Initialize Firebase
var config = {
	apiKey: "AIzaSyDxFq8QlAbDRIiQvSGD_a2C1Vwru0Q69rE",
	authDomain: "snail-heaven-1537271625768.firebaseapp.com",
	databaseURL: "https://snail-heaven-1537271625768.firebaseio.com",
	projectId: "snail-heaven-1537271625768",
	storageBucket: "snail-heaven-1537271625768.appspot.com",
	messagingSenderId: "888195818530"
};

export const firebaseApp = {
  fBase: firebase.initializeApp(config),
  fBaseDB:{}
};

const firebaseDB = firebaseApp.fBase.database();

firebaseApp.fBaseDB.setItem = (list) => {
	let amenity = {};
	for ( let i = 0 ; i < list.length ; i ++ ) {
		amenity[i] = {id: list[i].id, amenities: list[i].amenities }
	}
	// console.log(amenity);
	// firebaseApp.fBase.database().ref('amenity/').set({
 //    	id: id,
 //    	email: amenities,
	// });
}

firebaseApp.fBaseDB.getListing = callback => {
	firebaseDB.ref("listings/").once("value").then(snapshot => {
		if (callback) {
			callback(snapshot.val());
		}
	})
}

firebaseApp.fBaseDB.getData = (src, callback, childNode, equalValue) => {
	if ( childNode ) {
		firebaseDB.ref(src).orderByChild(childNode).equalTo(equalValue).once("value").then(snapshot => {
			if (callback) {
				callback(snapshot.val());
			}
		})
	} else {
		firebaseDB.ref(src).once("value").then(snapshot => {
			if (callback) {
				callback(snapshot.val());
			}
		})
	}
}

firebaseApp.fBaseDB.getDetailByID = ( ID ,callback ) => {
	firebaseDB.ref("details").orderByChild("id").equalTo(ID).once("value").then(snapshot => {
		if (callback) {
			callback(snapshot.val());
		}
	})
}


firebaseApp.fBaseDB.getListingByID = ( ID ,callback) => {
	firebaseDB.ref("listings").orderByChild("id").equalTo(ID).once("value").then(snapshot => {
		if (callback) {
			callback(snapshot.val());
		}
	})
}

firebaseApp.sortLatLng = ( data ) => {
	let location = [];
	for(let i = 0; i < data.length; i++ ) {
		let laAndLong = {lat:"", lng:""};
		laAndLong.lat = parseFloat(data[i].lat);
		laAndLong.lng = parseFloat(data[i].lng);
		location.push(laAndLong);
	}
	return location;
}

// firebaseApp.sortAmenity = () => {
// 	let amenitiesEN = ["Internet","Hot water","Air conditioning","Refrigerator","Laptop friendly workspace","Washer","Pets allowed","Kitchen","Gym","Elevator","Paid parking off premises","Free street parking"];
// 	let amenities = [[],[],[],[],[],[],[],[],[],[],[],[]];
// 	firebaseApp.fBaseDB.getData("amenity", (data)=>{
//     	for ( let i = 0 ; i < data.length ; i++ ) {
//     		for ( let j = 0 ; j < amenitiesEN.length ; j++ ) {
//     			let index = data[i].amenities.search(amenitiesEN[j]);
//     			if ( index !== -1 ) {
//     				amenities[j].push(data[i].id);
//     			}
//     		}
//     	}
//     	return amenities;
//     })
// }

firebaseApp.fBaseDB.getData1 = (src, callback, childNode, equalValue) => {
	if ( childNode ) {
		return firebaseDB.ref(src).orderByChild(childNode).equalTo(equalValue).once("value").then(snapshot => {
			return snapshot.val();
		})
	} else {
		return firebaseDB.ref(src).once("value").then(snapshot => {
			return snapshot.val();
			}
		)
	}
}

firebaseApp.sortAmenity = () => {
	return 	firebaseApp.fBaseDB.getData1("amenity").then((data)=>{
		// console.log(data);
		let amenitiesEN = ["Internet","Hot water","A/C","Refrigerator","Laptop friendly workspace","Washer","Pets allowed","Kitchen","Gym","Elevator","Paid parking off premises","Free street parking"];
		let amenities = [[],[],[],[],[],[],[],[],[],[],[],[]];
    	for ( let i = 0 ; i < data.length ; i++ ) {
    		for ( let j = 0 ; j < amenitiesEN.length ; j++ ) {
    			let index = data[i].amenities.search(amenitiesEN[j]);
    			if ( index !== -1 ) {
    				amenities[j].push(data[i].id);
    			}
    		}
    	}
    	return amenities;
	})
}	
