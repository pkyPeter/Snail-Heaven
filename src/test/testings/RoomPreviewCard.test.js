import React from "react";
import { shallow, mount, render } from "enzyme";
import { createSerializer } from "enzyme-to-json";
import renderer from "react-test-renderer";
import RoomPreviewCard from "../../../src/RoomPreviewCard.js";
import List from "../../../src/List.js";
import { firebaseApp } from "../../../src/firebaseApp.js";
import data  from "./testing_data/data.js";
expect.addSnapshotSerializer( createSerializer( { mode: "deep" } ) );

//test whether roomPreviewCard render correctly
jest.mock( "./imgs/airbnb.png", () => "airbnb.png" );


describe ( "RoomPreviewCard relevant tests", () => {
  it( "data is acquired from firebase correctly", done => {
    firebaseApp.fBaseDB.getDetailByIDWithPromise(data.realEstate.id)
    .then( ( dataFromFB ) => {
      // console.log(dataFromFB);
      expect(typeof dataFromFB).toBe("object");
      expect(dataFromFB["8251"].id).toBe(data.realEstate.id)
      expect(dataFromFB["8251"].price).toBe(data.realEstate.price + ".00");
      done();
    })
  })
	it( "roomPreviewCard renders correctly" , () => {
	  const wrapper = mount(
      <RoomPreviewCard 
      realEstate = {data.realEstate}
      resultAreaDisplayType = {data.resultAreaDisplayType}
      changeSelecteIndex = { () => {} }
      removeFromLoveList = { () => {} }
      putIntoLoveList = { ()=>{} }
      loveListStatus = { data.loveListStatus }
      getloveListStatusIndex={ ()=>{return 0} }
      openEmailForm = { ()=>{} }
      hideList = { ()=>{} }
      />
	  );
    let price = wrapper.find(".price").text();
    expect(price).toEqual("$2,212");
	  expect(wrapper).toMatchSnapshot();
	})
})
















// it( "List is rendered correctly", () => {
//   const wrapper = mount(
//     <List 
//       goLoveList= {false}
//       goLoveListPage={()=>{}}
//       completeList={realEstate}
//       loveListStatus={loveListStatus}
//       loveListDetail={[realEstate]}
//       openEmailForm={()=>{}}
//       selectedIndex={-1}
//       goPropertyPage={false}
//       changeSelecteIndex={()=>{}}
//       addSelectedIndex={()=>{}}
//       removeSelectedIndex={()=>{}}
//       putIntoLoveList={()=>{}}
//       removeFromLoveList={()=>{}}
//       getloveListStatusIndex={()=>{return 0;}}
//       currentViewData={realEstate}
//       filteredData={[realEstate]}
//       changeFilters={()=>{}}
//       filters={ filters }
//     />
//   );

//   // wrapper.instance().goSimpleDetail(realEstate.id, realEstate);
//   // wrapper.update();
//   // console.log(wrapper.state().toggleSimpleDetail)

//   // expect(wrapper.props().selectedIndex).toEqual(true);
//   // wrapper.setState({toggleSimpleDetail: true, currentSimpleDetail: realEstate})
//   // expect( wrapper ).toMatchSnapshot();
//  	  expect(wrapper).toMatchSnapshot();
// });
