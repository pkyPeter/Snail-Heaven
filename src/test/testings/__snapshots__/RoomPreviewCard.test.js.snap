// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomPreviewCard relevant tests roomPreviewCard renders correctly 1`] = `
<div
  className="results"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="airbnbContainer"
  >
    <img
      className="airbnb"
      src="airbnb.png"
    />
    Airbnb
  </div>
  <div
    className="img"
    style={
      Object {
        "backgroundImage": "url(https://a0.muscache.com/im/pictures/a0cf0ba1-d6...)",
      }
    }
  />
  <div
    className="description"
  >
    <div
      className="priceGesture absolute"
    >
      <div
        className="price"
      >
        $2,212
      </div>
      <div
        className="gesture"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-heart fa-w-16 icon"
          data-icon="heart"
          data-prefix="far"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-envelope fa-w-16 icon"
          data-icon="envelope"
          data-prefix="far"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 64H48C21.49 64 0 85.49 0 112v288c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V112c0-26.51-21.49-48-48-48zm0 48v40.805c-22.422 18.259-58.168 46.651-134.587 106.49-16.841 13.247-50.201 45.072-73.413 44.701-23.208.375-56.579-31.459-73.413-44.701C106.18 199.465 70.425 171.067 48 152.805V112h416zM48 400V214.398c22.914 18.251 55.409 43.862 104.938 82.646 21.857 17.205 60.134 55.186 103.062 54.955 42.717.231 80.509-37.199 103.053-54.947 49.528-38.783 82.032-64.401 104.947-82.653V400H48z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-thumbs-down fa-w-16 icon"
          data-icon="thumbs-down"
          data-prefix="far"
          onClick={[Function]}
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M466.27 225.31c4.674-22.647.864-44.538-8.99-62.99 2.958-23.868-4.021-48.565-17.34-66.99C438.986 39.423 404.117 0 327 0c-7 0-15 .01-22.22.01C201.195.01 168.997 40 128 40h-10.845c-5.64-4.975-13.042-8-21.155-8H32C14.327 32 0 46.327 0 64v240c0 17.673 14.327 32 32 32h64c11.842 0 22.175-6.438 27.708-16h7.052c19.146 16.953 46.013 60.653 68.76 83.4 13.667 13.667 10.153 108.6 71.76 108.6 57.58 0 95.27-31.936 95.27-104.73 0-18.41-3.93-33.73-8.85-46.54h36.48c48.602 0 85.82-41.565 85.82-85.58 0-19.15-4.96-34.99-13.73-49.84zM64 296c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zm330.18 16.73H290.19c0 37.82 28.36 55.37 28.36 94.54 0 23.75 0 56.73-47.27 56.73-18.91-18.91-9.46-66.18-37.82-94.54C206.9 342.89 167.28 272 138.92 272H128V85.83c53.611 0 100.001-37.82 171.64-37.82h37.82c35.512 0 60.82 17.12 53.12 65.9 15.2 8.16 26.5 36.44 13.94 57.57 21.581 20.384 18.699 51.065 5.21 65.62 9.45 0 22.36 18.91 22.27 37.81-.09 18.91-16.71 37.82-37.82 37.82z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <p>
      1
       間房間 · 
      1
       間廁所 · 
      整層住家
    </p>
    <p>
      大同區
    </p>
    <p
      className="updateTime"
    >
      3 days ago
    </p>
  </div>
</div>
`;
